In [1]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
import sklearn

**Q1.** Pandas version

What's the version of Pandas that you installed?

In [2]:
pd.__version__

'2.3.2'

In [3]:
df = pd.read_csv("car_fuel_efficiency.csv")

In [4]:
df.head()

engine_displacement  num_cylinders  horsepower  vehicle_weight  \
0                  170            3.0       159.0     3413.433759   
1                  130            5.0        97.0     3149.664934   
2                  170            NaN        78.0     3079.038997   
3                  220            4.0         NaN     2542.392402   
4                  210            1.0       140.0     3460.870990   

   acceleration  model_year  origin fuel_type         drivetrain  num_doors  \
0          17.7        2003  Europe  Gasoline    All-wheel drive        0.0   
1          17.8        2007     USA  Gasoline  Front-wheel drive        0.0   
2          15.1        2018  Europe  Gasoline  Front-wheel drive        0.0   
3          20.2        2009     USA    Diesel    All-wheel drive        2.0   
4          14.4        2009  Europe  Gasoline    All-wheel drive        2.0   

   fuel_efficiency_mpg  
0            13.231729  
1            13.688217  
2            14.246341  
3            16.912736  
4            12.488369

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9704 entries, 0 to 9703
Data columns (total 11 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   engine_displacement  9704 non-null   int64  
 1   num_cylinders        9222 non-null   float64
 2   horsepower           8996 non-null   float64
 3   vehicle_weight       9704 non-null   float64
 4   acceleration         8774 non-null   float64
 5   model_year           9704 non-null   int64  
 6   origin               9704 non-null   object 
 7   fuel_type            9704 non-null   object 
 8   drivetrain           9704 non-null   object 
 9   num_doors            9202 non-null   float64
 10  fuel_efficiency_mpg  9704 non-null   float64
dtypes: float64(6), int64(2), object(3)
memory usage: 834.1+ KB


In [6]:
df.index

RangeIndex(start=0, stop=9704, step=1)

**Q2.** Records count

How many records are in the dataset?

**9704** 

**Q3.** Fuel types 

How many fuel types are presented in the dataset?

2

In [7]:
df.nunique()

engine_displacement      36
num_cylinders            14
horsepower              192
vehicle_weight         9704
acceleration            162
model_year               24
origin                    3
fuel_type                 2
drivetrain                2
num_doors                 9
fuel_efficiency_mpg    9704
dtype: int64

**Q4. Missing values**

How many columns in the dataset have missing values?

**4**

In [8]:
df.isnull().sum()

engine_displacement      0
num_cylinders          482
horsepower             708
vehicle_weight           0
acceleration           930
model_year               0
origin                   0
fuel_type                0
drivetrain               0
num_doors              502
fuel_efficiency_mpg      0
dtype: int64

**Q5.** Max fuel efficiency

What's the maximum fuel efficiency of cars from Asia?

**23.75**


In [9]:
df[df['origin']=='Asia']['fuel_efficiency_mpg'].max()

23.759122836520497

Je peux aussi trouver la réponser avec Groupby

In [10]:
df.groupby('origin').fuel_efficiency_mpg.max()

origin
Asia      23.759123
Europe    25.967222
USA       24.971452
Name: fuel_efficiency_mpg, dtype: float64

**Q6. Median value of horsepower**

1) Find the median value of horsepower column in the dataset. **149.0**
2) Next, calculate the most frequent value of the same horsepower column. **152**
3) Use fillna method to fill the missing values in horsepower column with the most frequent value from the previous step.
4) Now, calculate the median value of horsepower once again. **152**

Has it changed?

Yes, it increased **it has increased**

Yes, it decreased

No



In [11]:
df.horsepower.median()

149.0

In [12]:
df.horsepower.mode()

0    152.0
Name: horsepower, dtype: float64

In [13]:
df['horsepower'] = df['horsepower'].fillna(152)

In [14]:
df['horsepower'].isnull()

0       False
1       False
2       False
3       False
4       False
        ...  
9699    False
9700    False
9701    False
9702    False
9703    False
Name: horsepower, Length: 9704, dtype: bool

In [15]:
df.horsepower.median()

152.0

In [16]:
df['horsepower'].nunique()

192

In [17]:
df['horsepower']

0       159.0
1        97.0
2        78.0
3       152.0
4       140.0
        ...  
9699    164.0
9700    154.0
9701    138.0
9702    177.0
9703    140.0
Name: horsepower, Length: 9704, dtype: float64

**Q7. Sum of weights**

1) Select all the cars from Asia
2) Select only columns vehicle_weight and model_year
3) Select the first 7 values
4) Get the underlying NumPy array. Let's call it X.
5) Compute matrix-matrix multiplication between the transpose of X and X. To get the transpose, use X.T. Let's call the result XTX.
6) Invert XTX.
7) Create an array y with values [1100, 1300, 800, 900, 1000, 1100, 1200].
8) Multiply the inverse of XTX with the transpose of X, and then multiply the result by y. Call the result w.
9) What's the sum of all the elements of the result?

In [18]:
df[df['origin']=='Asia']

engine_displacement  num_cylinders  horsepower  vehicle_weight  \
8                     250            1.0       174.0     2714.219310   
12                    320            5.0       145.0     2783.868974   
14                    200            6.0       160.0     3582.687368   
20                    150            3.0       197.0     2231.808142   
21                    160            4.0       133.0     2659.431451   
...                   ...            ...         ...             ...   
9688                  260            4.0       152.0     3948.404625   
9692                  180            3.0       188.0     3680.341381   
9693                  280            2.0       148.0     2545.070139   
9698                  180            1.0       131.0     3107.427820   
9703                  270            3.0       140.0     2908.043477   

      acceleration  model_year origin fuel_type         drivetrain  num_doors  \
8             10.3        2016   Asia    Diesel  Front-wheel drive       -1.0   
12            15.1        2010   Asia    Diesel    All-wheel drive        1.0   
14            14.9        2007   Asia    Diesel    All-wheel drive        0.0   
20            18.7        2011   Asia  Gasoline  Front-wheel drive        1.0   
21             NaN        2016   Asia  Gasoline  Front-wheel drive       -1.0   
...            ...         ...    ...       ...                ...        ...   
9688          15.5        2018   Asia    Diesel    All-wheel drive       -1.0   
9692          18.0        2016   Asia  Gasoline  Front-wheel drive        1.0   
9693          15.6        2012   Asia    Diesel    All-wheel drive        0.0   
9698          13.2        2005   Asia  Gasoline  Front-wheel drive       -2.0   
9703          14.7        2005   Asia    Diesel    All-wheel drive       -1.0   

      fuel_efficiency_mpg  
8               16.823554  
12              16.175820  
14              11.871091  
20              18.889083  
21              16.077730  
...                   ...  
9688            11.054830  
9692            11.711653  
9693            17.202782  
9698            13.933716  
9703            14.884467  

[3247 rows x 11 columns]

In [19]:
df_asia = df[df['origin']=='Asia'][['vehicle_weight','model_year']]

In [20]:
df_asia 

vehicle_weight  model_year
8        2714.219310        2016
12       2783.868974        2010
14       3582.687368        2007
20       2231.808142        2011
21       2659.431451        2016
...              ...         ...
9688     3948.404625        2018
9692     3680.341381        2016
9693     2545.070139        2012
9698     3107.427820        2005
9703     2908.043477        2005

[3247 rows x 2 columns]

In [21]:
df_asia.head(7)

vehicle_weight  model_year
8      2714.219310        2016
12     2783.868974        2010
14     3582.687368        2007
20     2231.808142        2011
21     2659.431451        2016
34     2844.227534        2014
38     3761.994038        2019

In [22]:
X = np.array(df_asia.head(7))

In [23]:
X

array([[2714.21930965, 2016.        ],
       [2783.86897424, 2010.        ],
       [3582.68736772, 2007.        ],
       [2231.8081416 , 2011.        ],
       [2659.43145076, 2016.        ],
       [2844.22753389, 2014.        ],
       [3761.99403819, 2019.        ]])

In [24]:
X.shape

(7, 2)

In [25]:
# Create the transpose of X
XT = X.T #we can use X.transpose() too

In [26]:
XT

array([[2714.21930965, 2783.86897424, 3582.68736772, 2231.8081416 ,
        2659.43145076, 2844.22753389, 3761.99403819],
       [2016.        , 2010.        , 2007.        , 2011.        ,
        2016.        , 2014.        , 2019.        ]])

In [27]:
XT.shape

(2, 7)

essayer de transposer X autrement

In [28]:
#Compute matrix-matrix multiplication between the transpose of X and X and call it XTX
XTX = XT.dot(X)

In [29]:
XTX

array([[62248334.33150762, 41431216.5073268 ],
       [41431216.5073268 , 28373339.        ]])

In [30]:
XTX.shape

(2, 2)

In [31]:
#Invert XTX
XTX_inv = np.linalg.inv(XTX)

In [32]:
XTX_inv

array([[ 5.71497081e-07, -8.34509443e-07],
       [-8.34509443e-07,  1.25380877e-06]])

In [33]:
#Create an array y with values [1100, 1300, 800, 900, 1000, 1100, 1200].
y = np.array([1100, 1300, 800, 900, 1000, 1100, 1200])

In [34]:
y

array([1100, 1300,  800,  900, 1000, 1100, 1200])

In [35]:
#Multiply the inverse of XTX with the transpose of X, and then multiply the result by y. Call the result w
XTX_invXT =XTX_inv.dot(XT)
w = XTX_invXT.dot(y)

Compute matrix-matrix multiplication between the transpose of X and X. To get the transpose, use X.T. Let's call the result XTX.

Invert XTX.

Create an array y with values [1100, 1300, 800, 900, 1000, 1100, 1200].

Multiply the inverse of XTX with the transpose of X, and then multiply the result by y. Call the result w.

What's the sum of all the elements of the result? **0.51**

In [36]:
w

array([0.01386421, 0.5049067 ])

***The details below are for my own understanding of how vector and matrix operations work

**Vector - vector multiplication**

Le resultat du vector vector multiplication est le résultat d'une addition, soit un chiffre

In [69]:
U = np.array([2, 4, 5, 6]).transpose()

In [70]:
V = np.array([1,0,0,2])

In [71]:
U.dot(V)

np.int64(14)

**Matrix - vector multiplication**

Renvoi un vector aussi large que le nombre de rangée de la matrice. Donc si la matrice a 3 rangées, le vector fruit de la multiplication matrice vector aura 3 chiffres, 3 rangées

In [79]:
U = np.array([
    [2, 4, 5, 6],
    [1, 2, 1, 2],
    [3, 1, 2, 1],
])

In [80]:
v = np.array([1, 0.5, 2, 1])

In [81]:
Uv = U.dot(V)

In [82]:
Uv.shape

(3,)

**Matrix - Matrix multiplication**

In [83]:
V = np.array([
    [1, 1, 2],
    [0, 0.5, 1],
    [0, 2, 1],
    [2, 1,0],
])

In [84]:
V

array([[1. , 1. , 2. ],
       [0. , 0.5, 1. ],
       [0. , 2. , 1. ],
       [2. , 1. , 0. ]])

In [85]:
U.dot(V)

array([[14. , 20. , 13. ],
       [ 5. ,  6. ,  5. ],
       [ 5. ,  8.5,  9. ]])